src/ratelimiter.rs

```diff
@@ -0,0 +1,140 @@
+
+        let needed = amount - total_bytes;
+        Duration::from_secs_f32(needed as f32 / self.bytes_per_second as f32)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_initial_state() {
+        let start = Instant::now();
+        let rl = RateLimiter::new(10, 10_000_000, start);
+
+        assert_eq!(rl.bytes_per_second, 10_000_000);
+        assert_eq!(rl.bytes_available(start), rl.bytes_per_second);
+    }
+
+    #[test]
+    fn test_bytes_available_after_one_second() {
+        let start = Instant::now();
+        let rl = RateLimiter::new(10, 10_000_000, start);
+
+        let now = start + Duration::from_secs(1);
+        assert_eq!(rl.bytes_available(now), 10_000_000);
+    }
+
+    #[test]
+    fn test_consume_bytes() {
+        let start = Instant::now();
+        let mut rl = RateLimiter::new(10, 10_000_000, start);
+
+        let now = start + Duration::from_secs(1);
+        assert_eq!(rl.bytes_available(now), 10_000_000);
+        rl.consume_bytes(now, 4_000_000);
+        assert_eq!(rl.available_bytes, 6_000_000);
+    }
+
+    #[test]
+    fn test_bytes_available_capped_at_max() {
+        let start = Instant::now();
+        let mut rl = RateLimiter::new(10, 10_000_000, start);
+
+        let now = start + Duration::from_secs(1);
+        rl.consume_bytes(now, 5_000_000);
+
+        let now = now + Duration::from_millis(500); // 0.5 seconds later
+        assert_eq!(rl.bytes_available(now), 10_000_000); // Should be capped at max
+
+        let now = now + Duration::from_millis(500); // 0.5 seconds later
+        assert_eq!(rl.bytes_available(now), 10_000_000); // Should be capped at max
+    }
+
+    #[test]
+    fn test_time_until_bytes_available() {
+        let start = Instant::now();
+        let mut rl = RateLimiter::new(10, 10_000_000, start);
+
+        let now = start + Duration::from_secs(1);
+        rl.consume_bytes(now, 9_000_000);
+        assert_eq!(rl.available_bytes, 1_000_000);
+
+        let wait_time = rl.time_until_bytes_available(now, 9_000_000);
+        // at 10MB/s, 800ms for 800KB
+        assert!(wait_time > Duration::from_millis(799) && wait_time < Duration::from_millis(801));
+    }
+
+    #[test]
+    fn test_immediate_availability() {
+        let start = Instant::now();
+        let mut rl = RateLimiter::new(10, 10_000_000, start);
+
+        let now = start + Duration::from_secs(1);
+        rl.consume_bytes(now, 9_000_000);
+
+        assert_eq!(
+            rl.time_until_bytes_available(now, 1_000_000),
+            Duration::from_secs(0)
+        );
+    }
+
+    #[test]
+    fn test_long_wait_time() {
+        let start = Instant::now();
+        let mut rl = RateLimiter::new(10, 10_000_000, start);
+
+        let now = start + Duration::from_secs(1);
+        rl.consume_bytes(now, 9_000_000);
+
+        // this is not true, there will never be 20M available in the bucket.
+        // not sure if this case should throw when asking for > bps
```

Comment:
We could make `time_until_bytes_available` return an `Option<Duration>`, where `None` means “you requested more than the bucket’s capacity”, but then at the call site it would be a programming error anyway, hmm. We could put in an assert that the request is not larger than the capacity, probably it’s probably better to crash the program than to wait indefinitely.